src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala

```diff
@@ -226,38 +234,47 @@ final case class PurposesApiServiceImpl(
     }
   }
 
+  def getCurrentVersion(purpose: PurposeProcess.Purpose): Option[PurposeProcess.PurposeVersion] =
+    purpose.versions
+      .filter(v => v.state != PurposeProcess.PurposeVersionState.WAITING_FOR_APPROVAL)
+      .sortBy(_.createdAt)
+      .lastOption
+
+  def getWaitingForApproval(purpose: PurposeProcess.Purpose): Option[PurposeProcess.PurposeVersion] =
+    purpose.versions.find(_.state == PurposeProcess.PurposeVersionState.WAITING_FOR_APPROVAL)
+
   private def enhancePurpose(
     purpose: PurposeProcess.Purpose,
     eServices: Seq[CatalogProcess.EService],
     producers: Seq[TenantProcess.Tenant],
-    consumers: Seq[TenantProcess.Tenant]
-  ): Future[Purpose] = for {
-    eService <- eServices.find(_.id == purpose.eserviceId).toFuture(EServiceNotFound(purpose.eserviceId))
-    producer <- producers.find(_.id == eService.producerId).toFuture(TenantNotFound(eService.producerId))
-    consumer <- consumers.find(_.id == purpose.consumerId).toFuture(TenantNotFound(purpose.consumerId))
-    currentVersion            = purpose.versions
-      .filter(v => v.state != PurposeProcess.PurposeVersionState.WAITING_FOR_APPROVAL)
-      .sortBy(_.createdAt)
-      .lastOption
-    waitingForApprovalVersion = purpose.versions.find(
-      _.state == PurposeProcess.PurposeVersionState.WAITING_FOR_APPROVAL
+    consumers: Seq[TenantProcess.Tenant],
+    requesterId: UUID
+  )(implicit context: Seq[(String, String)]): Future[Purpose] = for {
+    eService          <- eServices.find(_.id == purpose.eserviceId).toFuture(EServiceNotFound(purpose.eserviceId))
+    producer          <- producers.find(_.id == eService.producerId).toFuture(TenantNotFound(eService.producerId))
+    consumer          <- consumers.find(_.id == purpose.consumerId).toFuture(TenantNotFound(purpose.consumerId))
+    agreement         <- getLatestAgreement(requesterId, eService, ec, agreementProcessService).flatMap(
+      _.toFuture(AgreementNotFound(requesterId))
```

Comment:
Not at the moment, but you may possibly request a purpose of another organization, we must be sure to retrieve the correct agreement
```suggestion
    agreement         <- getLatestAgreement(purpose.consumerId, eService, ec, agreementProcessService).flatMap(
      _.toFuture(AgreementNotFound(purpose.consumerId))
```